httpbp/client_middlewares.go

```diff
@@ -0,0 +1,199 @@
+				resp, err := next.RoundTrip(req)
+				if err != nil {
+					return nil, err
+				}
+				// circuit break on any HTTP 5xx code
+				if resp.StatusCode >= http.StatusInternalServerError {
+					// read & close to ensure underlying RoundTripper
+					// (http.Transport) is able to re-use the persistent TCP
+					// connection.
+					_, _ = io.ReadAll(resp.Body)
+					err = resp.Body.Close()
+					if err != nil {
+						return nil, err
+					}
+					return nil, err
+				}
+				return resp, nil
+			})
+			if err != nil {
+				return nil, err
+			}
+			return resp.(*http.Response), nil
+		})
+	}
+}
+
+func applyMiddlewares(transport http.RoundTripper, middlewares ...ClientMiddleware) http.RoundTripper {
+	// add middlewares in reverse so the first in the list is the outermost
+	for i := len(middlewares) - 1; i >= 0; i-- {
+		transport = middlewares[i](transport)
+	}
+	return transport
+}
+
+// Retries provides a retry middleware.
+func Retries(retryOptions ...retry.Option) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+
+			if len(retryOptions) == 0 {
+				retryOptions = []retry.Option{retry.Attempts(1)}
+			}
+
+			retrybp.Do(req.Context(), func() error {
+				resp, err = next.RoundTrip(req)
+				if err != nil {
+					return err
+				}
+				return nil
+			}, retryOptions...)
+			if err != nil {
+				return nil, err
+			}
+			return resp, nil
+		})
+	}
+}
+
+// MaxConcurrency is a middleware to ensure that there will be only a maximum
+// number of requests concurrently in-flight at any given time.
+func MaxConcurrency(config ClientConfig) ClientMiddleware {
+	var (
+		activeRequests int64
+	)
+	maxConcurrency := int64(config.MaxConcurrency)
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+			attemptedRequests := atomic.AddInt64(&activeRequests, 1)
+			defer atomic.AddInt64(&activeRequests, -1)
+
+			if maxConcurrency > 0 && attemptedRequests > maxConcurrency {
+				return nil, ErrConcurrencyLimit
+			}
+			return next.RoundTrip(req)
+		})
+	}
+}
+
+// MonitorClient is a HTTP client middleware that wraps HTTP requests in a
+// client span.
+func MonitorClient(config ClientConfig) ClientMiddleware {
+	slug := config.Slug + ".request"
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+			span, ctx := opentracing.StartSpanFromContext(
+				req.Context(),
+				slug,
+				tracing.SpanTypeOption{Type: tracing.SpanTypeClient},
+			)
+			span.SetTag("http.method", req.Method)
+			span.SetTag("http.url", req.URL)
+
+			defer func() {
+				span.FinishWithOptions(tracing.FinishOptions{
+					Ctx: req.Context(),
+					Err: err,
+				}.Convert())
+			}()
+			req = req.WithContext(ctx)
```

Comment:
Overwriting locals is enough of a cognitive burden that I think it's worth just inlining this on the following line